### PR TITLE
[BandcampWeekly] Add new extractor

### PR DESCRIPTION
**Pull request by adamvoss**
### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [X] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The Bandcamp Weekly is a weekly audio show accessible via the Bandcamp homepage that showcases audio tracks and interviews with Bandcamp artists.  This show is not covered by the existing Bandcamp extractors.  This PR adds the ability to download episodes of the Bandcamp Weekly.

### CODE REVIEW

**Comment by vossad01** [comment]
Near as I can tell, the Travis CI failure is not specific to this PR at this point.  The only Bandcamp related failure is the following and appears to have been preexisting:

> test_Bandcamp_1 (test.test_download.TestDownload): ... FAIL


If there was a different/better way to handle the RegEx conflict between this and the existing BandcampAlbum extractor, I would be interested, this was the best solution I could see with my understanding of the architecture.

**Comment by dstftw** [comment]
```
PS C:\dev\youtube-dl\master> py -3.6 .\test\test_all_urls.py
..F..............
======================================================================
FAIL: test_no_duplicated_ie_names (__main__.TestAllURLsMatching)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".\test\test_all_urls.py", line 139, in test_no_duplicated_ie_names
    'Multiple extractors with the same IE_NAME "%s" (%s)' % (ie_name, ', '.join(ie_list)))
AssertionError: 2 != 1 : Multiple extractors with the same IE_NAME "bandcamp:album" (BandcampAlbumIE, BandcampArtistAlbumIE)

----------------------------------------------------------------------
Ran 17 tests in 9.014s

FAILED (failures=1)
```

**Comment by vossad01** [comment]
Thanks!  Correction pushed:

```
$ python2 ./test/test_all_urls.py
.................
----------------------------------------------------------------------
Ran 17 tests in 10.617s

OK
```

File: youtube_dl/extractor/bandcamp.py [hunk 1]
```diff
@@ -248,3 +241,78 @@ def _real_extract(self, url):
             'title': title,
             'entries': entries,
         }
+
+
+class BandcampArtistDefaultAlbumIE(BandcampAlbumIE):
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: What's the point of this extractor? It's covered by album extractor. Remove.

File: youtube_dl/extractor/bandcamp.py [hunk 2]
```diff
@@ -248,3 +241,78 @@ def _real_extract(self, url):
...
+            self._search_regex(
+                r'data-blob=(["\'])(?P<blob>{.+?})\1', webpage,
+                'blob', group='blob'),
+            video_id, transform_source=unescapeHTML)
+
+        show = blob['bcw_show']
+
+        # This is desired because any invalid show id redirects to `bandcamp.com`
+        # which happens to expose the latest Bandcamp Weekly episode.
+        video_id = str(show['show_id'])
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `compat_str`.

File: youtube_dl/extractor/bandcamp.py [hunk 3]
```diff
@@ -248,3 +241,78 @@ def _real_extract(self, url):
...
+
+        def to_format_dictionaries(audio_stream):
+            dictionaries = [{'format_id': kvp[0], 'url': kvp[1]} for kvp in audio_stream.items()]
+
+            for dictionary in dictionaries:
+                if 'mp3' in dictionary['format_id']:
+                    dictionary['ext'] = 'mp3'
+                if 'opus' in dictionary['format_id']:
+                    dictionary['ext'] = 'opus'
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Code duplication.

File: youtube_dl/extractor/bandcamp.py (hunk 1 above)
Comment (adamvoss) [inline youtube_dl/extractor/bandcamp.py]:
```python
$ python2
> album_VALID_URL = r'https?://(?:(?P<subdomain>[^.]+)\.)?bandcamp\.com/album/(?:(?P<album_id>[^?#]+)|/?(?:$|[?#]))'
> artist_default_album_VALID_URL = r'https?://(?P<subdomain>[^.]+)\.bandcamp\.com/?(?:$|[?#])(?P<album_id>)'
> bool(re.match(album_VALID_URL, 'http://dotscale.bandcamp.com'))
False
> bool(re.match(artist_default_album_VALID_URL, 'http://dotscale.bandcamp.com'))
True
```

The extractor for the album extractor had to be made less general so it did not clobber the Bandcamp Weekly extractor.  However, the resulting regex would not capture cases such as 'http://dotscale.bandcamp.com' where an album is displayed on an "artists" page.  Thus this extractor to preserve capturing this case.

File: youtube_dl/extractor/bandcamp.py (hunk 3 above)
Comment (adamvoss) [inline youtube_dl/extractor/bandcamp.py]: @dstftw Can I get a hint here?  Are you saying you want something like this?

```python
def set_extension(ext):
    if ext in dictionary['format_id']:
        dictionary['ext'] = ext

set_extension('mp3')
set_extension('opus')
```

File: youtube_dl/extractor/bandcamp.py (hunk 1 above)
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: No. Override `suitable`.

File: youtube_dl/extractor/bandcamp.py (hunk 3 above)
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: No.
```
for ext in ('mp3', 'opus')  # or KNOWN_EXTENSIONS:
  if ext in d...:
    d... = ext
    break
```

**Comment by adamvoss** [comment]
Thank your for reviewing this and providing clarification on the needed changes!  The requested changes have been made.

File: youtube_dl/extractor/bandcamp.py [hunk 4]
```diff
@@ -153,8 +154,10 @@ def _real_extract(self, url):
 
 class BandcampAlbumIE(InfoExtractor):
     IE_NAME = 'Bandcamp:album'
-    _VALID_URL = r'https?://(?:(?P<subdomain>[^.]+)\.)?bandcamp\.com(?:/album/(?P<album_id>[^?#]+)|/?(?:$|[?#]))'
-
+    VALID_URLS = \
+    [r'https?://(?:(?P<subdomain>[^.]+)\.)?bandcamp\.com/album/(?:(?P<album_id>[^?#]+)|/?(?:$|[?#]))',
+        r'https?://(?P<subdomain>[^.]+)\.bandcamp\.com/?(?:$|[?#])(?P<album_id>)']
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: What are you trying to do here? There should be a single regex.

File: youtube_dl/extractor/bandcamp.py [hunk 5]
```diff
@@ -248,3 +261,62 @@ def _real_extract(self, url):
...
+        return {
+            'id': video_id,
+            'title': show['audio_title'] + ': ' + show['subtitle'],
+            'description': show.get('desc'),
+            'duration': show.get('audio_duration'),
+            'webpage_url': 'https://bandcamp.com/?show=' + video_id,
+            'is_live': False,
+            'release_date': unified_strdate(show.get('published_date')),
+            'series': 'Bandcamp Weekly',
+            'episode_id': compat_str(video_id),
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `episode_number`.

File: youtube_dl/extractor/bandcamp.py [hunk 6]
```diff
@@ -248,3 +261,62 @@ def _real_extract(self, url):
             'title': title,
             'entries': entries,
         }
+
+
+class BandcampWeeklyIE(InfoExtractor):
+    IE_NAME = 'Bandcamp:bandcamp_weekly'
+    _VALID_URL = r'https?://(?:www\.)?bandcamp\.com/\?(?:\w+=\w+&)*show=(?P<id>\d+)(?:&\w+=\w+)*#?'
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `(?:&\w+=\w+)*#?` is noop. Does not match `https://bandcamp.com/?blah@&show=224`.

File: youtube_dl/extractor/bandcamp.py [hunk 7]
```diff
@@ -248,3 +261,62 @@ def _real_extract(self, url):
...
+            return dictionaries
+
+        formats = to_format_dictionaries(show['audio_stream'])
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': show['audio_title'] + ': ' + show['subtitle'],
+            'description': show.get('desc'),
+            'duration': show.get('audio_duration'),
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: `float_or_none`.

File: youtube_dl/extractor/bandcamp.py (hunk 4 above)
Comment (adamvoss) [inline youtube_dl/extractor/bandcamp.py]: [See comments on previous revision](https://github.com/rg3/youtube-dl/pull/12758#discussion_r118849945).  This originally started with 79b3f1830328117edf9e301a42fd8cb692eaac0f.

It sounds like instead for overriding `suitable` you want a big regex with an alternation (`|`) which will require renaming the subgroups to be unique and either later on.

File: youtube_dl/extractor/bandcamp.py (hunk 5 above)
Comment (adamvoss) [inline youtube_dl/extractor/bandcamp.py]: As it happens, the `video_id` does not match the episode number.

For example,
Episode 1 is 1, http://bandcamp.com/?show=1
Episode 2 is 6, http://bandcamp.com/?show=6

File: youtube_dl/extractor/bandcamp.py (hunk 4 above)
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: [No, you don't get it](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/extractor/acast.py#L80).

**Comment by adamvoss** [comment]
Another round of revisions complete

File: youtube_dl/extractor/bandcamp.py [hunk 8]
```diff
@@ -248,3 +253,71 @@ def _real_extract(self, url):
...
+        }
+    }, {
+        'url': 'https://bandcamp.com/?blah@&show=228',
+        'md5': '894ef5e83d7f653d99f644c5410be643',
+        'info_dict': {
+            'id': '228',
+            'ext': 'opus',
+            'title': 'BC Weekly May 2nd 2017: Shifting Gears',
+            'description': 'Featuring an interview with Malcolm Catto of The Heliocentrics, plus new music by Shabazz Palaces, Jane Weaver, and Morgan Guerin.',
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: All duplicate tests should be set `only_matching`.

Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: All duplicate tests should be set `only_matching`.

File: youtube_dl/extractor/bandcamp.py [hunk 9]
```diff
@@ -248,3 +253,71 @@ def _real_extract(self, url):
             'title': title,
             'entries': entries,
         }
+
+
+class BandcampWeeklyIE(InfoExtractor):
+    IE_NAME = 'Bandcamp:bandcamp_weekly'
+    _VALID_URL = r'https?://(?:www\.)?bandcamp\.com/?\?(?:[^/]*&)?show=(?P<id>\d+)(?:$|[&#])'
```
Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Does not match https://bandcamp.com/?blah/blah&show=228.

File: youtube_dl/extractor/bandcamp.py (hunk 8 above)
Comment (adamvoss) [inline youtube_dl/extractor/bandcamp.py]: If this is testing a different episode/show is it still considered a duplicate?

Comment (dstftw) [inline youtube_dl/extractor/bandcamp.py]: Yes, if it tests the same extraction scenario. There is only one extraction scenario in this code.

File: youtube_dl/extractor/bandcamp.py (hunk 9 above)
Comment (adamvoss) [inline youtube_dl/extractor/bandcamp.py]: I hadn't realized that was valid.  Fixed, thanks!

**Comment by dstftw** [comment]
Squash commits.

**Comment by adamvoss** [comment]
Squashed and rebased.

**Pull request closed without merging**